Introduce cursor tracking akin to jdbc input (#205)
logstash-plugins/logstash-input-elasticsearch: docs/index.asciidoc

NOTE: experimental:[] `tracking_field` and related settings are experimental and subject to change in the future

It is sometimes desirable to track the value of a particular field between two jobs:
* avoid re-processing the entire result set of a long query after an unplanned restart
* only grab new data from an index instead of processing the entire set on each job

For this, the Elasticsearch input plugin provides the <<plugins-{type}s-{plugin}-tracking_field>> and <<plugins-{type}s-{plugin}-tracking_field_seed>> options.
When <<plugins-{type}s-{plugin}-tracking_field>> is set, the plugin will record the value of that field for the last document retrieved in a run into
a file (location defaults to <<plugins-{type}s-{plugin}-last_run_metadata_path>>).

The user can then inject this value in the query using the placeholder `:last_value`. The value will be injected into the query
before execution, and the updated after the query completes, assuming new data was found.

This feature works best when:
* the query sorts by the tracking field
* the field type has enough resolution so that two events are unlikely to have the same value for the field

> The plugin also offers another placeholder called `:present` used to inject the nano-second based value of "now-30s".

Reviewer: Would it make sense to move this placeholder description below the suggestion and introduce it with something along the lines of:

"To support this, the plugin also offers a placeholder called..."

 - When the hit result contains top-level fields that are {logstash-ref}/processing.html#reserved-fields[reserved in Logstash] but do not have the expected shape. Use the <<plugins-{type}s-{plugin}-target>> directive to avoid conflicts with the top-level namespace.
 - When <<plugins-{type}s-{plugin}-docinfo>> is enabled and the docinfo fields cannot be merged into the hit result. Combine <<plugins-{type}s-{plugin}-target>> and <<plugins-{type}s-{plugin}-docinfo_target>> to avoid conflict.

[id="plugins-{type}s-{plugin}-cursor"]
==== Tracking a field's value across runs

NOTE: experimental:[] `tracking_field` and related settings are experimental and subject to change in the future

> 

Reviewer 2: ```suggestion
.Technical Preview: Tracking a field's value
****
The feature that allows tracking a field's value across runs is in _Technical Preview_.
Configuration options and implementation details are subject to change in minor releases without being preceded by deprecation warnings.
****

```

Reviewer 2: We don't need both notes, so please decide which treatment and working you like.

> It is sometimes desirable to track the value of a particular field between two jobs:
> * avoid re-processing the entire result set of a long query after an unplanned restart
> * only grab new data from an index instead of processing the entire set on each job

Reviewer 2: ```suggestion
Some uses cases require tracking the value of a particular field between two jobs.
Examples include:

* avoiding the need to re-process the entire result set of a long query after an unplanned restart
* grabbing only new data from an index instead of processing the entire set on each job.
```

> For this, the Elasticsearch input plugin provides the <<plugins-{type}s-{plugin}-tracking_field>> and <<plugins-{type}s-{plugin}-tracking_field_seed>> options.
> When <<plugins-{type}s-{plugin}-tracking_field>> is set, the plugin will record the value of that field for the last document retrieved in a run into
> a file (location defaults to <<plugins-{type}s-{plugin}-last_run_metadata_path>>).

Reviewer 2: ```suggestion
The Elasticsearch input plugin provides the <<plugins-{type}s-{plugin}-tracking_field>> and <<plugins-{type}s-{plugin}-tracking_field_seed>> options.
When <<plugins-{type}s-{plugin}-tracking_field>> is set, the plugin records the value of that field for the last document retrieved in a run into
a file. 
(The file location defaults to <<plugins-{type}s-{plugin}-last_run_metadata_path>>.)
```

> The user can then inject this value in the query using the placeholder `:last_value`. The value will be injected into the query
> before execution, and the updated after the query completes if new data was found.

Reviewer 2: ```suggestion
You can then inject this value in the query using the placeholder `:last_value`. 
The value will be injected into the query before execution, and then updated after the query completes if new data was found.
```

This feature works best when:

> . the query sorts by the tracking field;
> . the timestamp field is added by {es};
> . the field type has enough resolution so that two events are unlikely to have the same value.

Reviewer 2: ```suggestion
* the query sorts by the tracking field,
* the timestamp field is added by {es}, and
* the field type has enough resolution so that two events are unlikely to have the same value.
```

It is recommended to use a tracking field whose type is https://www.elastic.co/guide/en/elasticsearch/reference/current/date_nanos.html[date nanoseconds].
If the tracking field is of this data type, an extra placeholder called `:present` can be used to inject the nano-second based value of "now-30s".
This placeholder is useful as the right-hand side of a range filter, allowing the collection of
new data but leaving partially-searcheable bulk request data to the next scheduled job.

> Below is a series of steps to help set up the "tailing" of data being written to a set of indices, using a date nanosecond field
> added by an Elasticsearch ingest pipeline, and the `tracking_field` capability of this plugin.

Reviewer 2: ```suggestion
id="plugins-{type}s-{plugin}-tracking-sample"]
===== Sample configuration: Track field value across runs

This section contains a series of steps to help you set up the "tailing" of data being written to a set of indices, using a date nanosecond field added by an Elasticsearch ingest pipeline and the `tracking_field` capability of this plugin.
```

When <<plugins-{type}s-{plugin}-tracking_field>> is set, the plugin records the value of that field for the last document retrieved in a run into
a file. 
(The file location defaults to <<plugins-{type}s-{plugin}-last_run_metadata_path>>.)

You can then inject this value in the query using the placeholder `:last_value`. 
The value will be injected into the query before execution, and then updated after the query completes if new data was found.

This feature works best when:

* the query sorts by the tracking field,
* the timestamp field is added by {es}, and
* the field type has enough resolution so that two events are unlikely to have the same value.

Consider using a tracking field whose type is https://www.elastic.co/guide/en/elasticsearch/reference/current/date_nanos.html[date nanoseconds].
If the tracking field is of this data type, you can use an extra placeholder called `:present` to inject the nano-second based value of "now-30s".
This placeholder is useful as the right-hand side of a range filter, allowing the collection of
new data but leaving partially-searchable bulk request data to the next scheduled job.

> id="plugins-{type}s-{plugin}-tracking-sample"]

Reviewer 2: ```suggestion
[id="plugins-{type}s-{plugin}-tracking-sample"]
```

Reviewer 2: Formatting error.  This one's on me.

 - When the hit result contains top-level fields that are {logstash-ref}/processing.html#reserved-fields[reserved in Logstash] but do not have the expected shape. Use the <<plugins-{type}s-{plugin}-target>> directive to avoid conflicts with the top-level namespace.
 - When <<plugins-{type}s-{plugin}-docinfo>> is enabled and the docinfo fields cannot be merged into the hit result. Combine <<plugins-{type}s-{plugin}-target>> and <<plugins-{type}s-{plugin}-docinfo_target>> to avoid conflict.

[id="plugins-{type}s-{plugin}-cursor"]
==== Tracking a field's value across runs

> 
> NOTE: experimental:[] `tracking_field` and related settings are experimental and subject to change in the future
> .Technical Preview: Tracking a field's value

Author: ```suggestion

.Technical Preview: Tracking a field's value
```

There is no schedule by default. If no schedule is given, then the statement is run
exactly once.

[id="plugins-{type}s-{plugin}-schedule_overlap"]
===== `schedule_overlap`

  * Value type is <<boolean,boolean>>
  * Default value is `true`

Whether to allow queuing of a scheduled run if a run is occurring.

> While this is ideal for ensuring a new run happens immediatelly after the previous on finishes if there

Reviewer: ```suggestion
While this is ideal for ensuring a new run happens immediately after the previous on finishes if there
```

This option can be useful to avoid populating unknown fields when a downstream schema such as ECS is enforced.
It is also possible to target an entry in the event's metadata, which will be available during event processing but not exported to your outputs (e.g., `target \=> "[@metadata][_source]"`).

[id="plugins-{type}s-{plugin}-tracking_field"]
===== `tracking_field`

* Value type is <<string,string>>
* There is no default value for this setting.

Which field from the last event of a previous run will be used a cursor value for the following run.
The value of this field is injected into each query if the query uses the placeholder `:last_value`.
For the first query after a pipeline is started, the value used is either read from <<plugins-{type}s-{plugin}-last_run_metadata_path>> file,
or taken from <<plugins-{type}s-{plugin}-tracking_field_seed>> setting.

> Note: The tracking value is updated only after the PIT+search_after run completes, it won't update during the search_after pagination. This is to allow use of slices.

Author: ```suggestion
Note: The tracking value is updated after each page is read and at the end of each Point in Time. In case of a crash the last saved value will be used so some duplication of data can occur. For this reason the use of unique document IDs for each event is recommended in the downstream destination.
```

It is also possible to target an entry in the event's metadata, which will be available during event processing but not exported to your outputs (e.g., `target \=> "[@metadata][_source]"`).

[id="plugins-{type}s-{plugin}-tracking_field"]
===== `tracking_field`

* Value type is <<string,string>>
* There is no default value for this setting.

Which field from the last event of a previous run will be used a cursor value for the following run.
The value of this field is injected into each query if the query uses the placeholder `:last_value`.
For the first query after a pipeline is started, the value used is either read from <<plugins-{type}s-{plugin}-last_run_metadata_path>> file,
or taken from <<plugins-{type}s-{plugin}-tracking_field_seed>> setting.

Note: The tracking value is updated after each page is read and at the end of each Point in Time. In case of a crash the last saved value will be used so some duplication of data can occur. For this reason the use of unique document IDs for each event is recommended in the downstream destination.

[id="plugins-{type}s-{plugin}-tracking_field_seed"]
===== `tracking_field_seed`

* Value type is <<string,string>>

> * There is no default value for this setting.

Reviewer: Looking at the code it defaults to `Time.now`?

The starting value for the <<plugins-{type}s-{plugin}-tracking_field>> if there is no <<plugins-{type}s-{plugin}-last_run_metadata_path>> already.

> For a nanosecond timestamp based field, a suggested seed value something like `1980-01-01T23:59:59.999999999Z`.

Reviewer: Maybe an explanation of behavior when a `tracking_field` is chosen without a `tracking_field_seed`?

Author: no need now that seed has a default to epoch.
